### PR TITLE
Chapter 9: Forms & Chapter 10: Authentication

### PR DESCRIPTION
**Pull request by schintees**

### CODE REVIEW

File: src/app/app.component.ts [hunk 1]
```diff
@@ -7,4 +7,9 @@ import { Component } from '@angular/core';
 })
 export class AppComponent {
   title = 'online-shop-angular';
+
+  @HostListener("window:beforeunload")
```
Comment (antonio-marasescu-duran) [inline src/app/app.component.ts]: **question:** why did you add this? This will logout you after each refresh, you don't really want to do this.

File: src/app/components/containers/products-details/products-details.component.html [hunk 2]
```diff
@@ -1,8 +1,10 @@
 <div *ngIf="product$ | async as product">
   <app-products-details-view
     [product]="product"
+    [isAdmin]="(isAdmin$ | async) || undefined"
```
Comment (antonio-marasescu-duran) [inline src/app/components/containers/products-details/products-details.component.html]: **suggestion(non-blocking):** if an object does not exist, by whatever reason, it will be undefined anyhow

File: src/app/components/containers/products-form/products-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    this.isAddMode ? this.createProduct() : this.updateProduct();
+  }
+
+  onCancel() {
+    this.navigationService.navigateToProductsPage()
+  }
+
+  private createProduct() {
+    this.productService.addProduct(this.productForm.value)
```
Comment (antonio-marasescu-duran) [inline src/app/components/containers/products-form/products-form.component.ts:61]: **suggestion(non-blocking):** As a personal preference, there are times where you truly use an observable (as a long-term value) otherwise you truly only need a Promise. My preference in case where you fire a request and only await for 1 response is to transform the observable into a Promise using `firstValueFrom`, making the function async and using await.

File: src/app/app.component.ts (hunk 1 above)
Comment (schintees) [reply src/app/app.component.ts]: I added it because in the description we have this: _Automatically redirect the user to the login view each time he opens or refreshes the application_

File: src/app/interceptors/auth-interceptor.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    constructor(private auth: AuthService) { }
+
+    intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+        request = request.clone({
+            setHeaders: {
+                Authorization: `Bearer ${this.auth.getToken()}`
+            }
+        });
+        return next.handle(request);
```
Comment (antonio-marasescu-duran) [inline src/app/interceptors/auth-interceptor.ts]: **suggestion:** Maybe you can handle the case when the response is a 401 (the token expired)

File: src/app/modules/user/components/containers/user-login/user-login.component.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    this.userService.login(this.loginForm.value)
+      .pipe(untilDestroyed(this))
+      .pipe(catchError(error => this.catchAuthError(error)))
+      .subscribe(() => {
+        this.navigationService.navigateToProductsPage()
+      });
+
+  }
+
+  catchAuthError(error: any): Observable<any> {
```
Comment (antonio-marasescu-duran) [inline src/app/modules/user/components/containers/user-login/user-login.component.ts]: **suggestion(non-blocking):** try to avoid `any`, does the error truly not have any type?

File: src/app/modules/user/components/containers/user-navigation/user-navigation.component.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  onLogin() {
+    this.navigationService.navigateToLoginPage();
+  }
+
+  onHome() {
+    this.navigationService.navigateToProductsPage();
+  }
+
+  onAddProduct() {
+    this.navigationService.navigateToAddProducttPage();
```
Comment (antonio-marasescu-duran) [inline src/app/modules/user/components/containers/user-navigation/user-navigation.component.ts]: **chore(non-blocking):** typo on `navigateToAddProducttPage`, double `t`

File: src/app/modules/user/services/auth.service.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      );
+  }
+
+  logout() {
+    localStorage.clear();
+    this.currentUserSubject$.next(undefined);
+  }
+
+  isLoggedIn() {
+    return this.getToken() ? true : false;
```
Comment (antonio-marasescu-duran) [inline src/app/modules/user/services/auth.service.ts]: **suggestion(non-blocking):** this can be simplified to `!!this.getToken()`, also to keep in mind the JWT token usually has an expiration period (do not handle this here tho). It is in most normal case (again not here) composed from an access_token, id_token and refresh_token. If you have an interest to read: https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/

File: src/app/modules/user/user.module.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      provide: HTTP_INTERCEPTORS,
+      useClass: AuthInterceptor,
+      multi: true
+    }
+  ],
+  exports: [
+    UserNavigationComponent
+  ]
+})
+export class UserModule { }
```
Comment (antonio-marasescu-duran) [inline src/app/modules/user/user.module.ts:42]: **compliment:** i see you added another module, though not required by the training, nice work

File: src/styles/mixins.scss [hunk 9]
```diff
@@ -20,3 +20,32 @@
...
+        justify-content: center;
+        margin: 100px 0px;
+    }
+
+    mat-form-field {
+        width: 100%;
+        min-width: 300px;
+    }
+
+    mat-card-title,
```
Comment (antonio-marasescu-duran) [inline src/styles/mixins.scss:36]: **suggestion(non-blocking):** try to avoid as much as possible modifing directly angular material classes. First try, modifing through the theme. Next best scenario is to create wrapper components for the specific material components you want to modify and have it tested in isolation (maybe a Component Library with Storybook). Final scenario is to modify them through the scss, but keep track of the variables to keep it consistent. Anyhow, ignoring my talk :)) don't change here anything, just to keep in mind

**Review by antonio-marasescu-duran — CHANGES_REQUESTED** [review]
Left some comments

File: src/app/app.component.ts (hunk 1 above)
Comment (antonio-marasescu-duran) [reply src/app/app.component.ts]: Aaa, I forgot about that, it was an old description before we had JWT for the training, sorry. I will update the training

**Review by antonio-marasescu-duran — APPROVED** [review]
